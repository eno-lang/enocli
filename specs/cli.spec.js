const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

const DOCUMENT_PATH = path.join(__dirname, 'fixtures/document.eno');
const OUTPUT_PATH = path.join(__dirname, 'temp_output/cli_output.js');

describe('CLI', () => {
  describe('generates a parser given valid arguments', () => {
    let cliOutput = [];
    let generatedCode;

    beforeAll(async () => {
      await new Promise(resolve => {
        const cli = spawn('node', ['bin/enocli.js', DOCUMENT_PATH, `javascript=${OUTPUT_PATH}`]);

        cli.stdout.on('data', data => cliOutput.push({ stdout: data.toString() }));
        cli.stderr.on('data', data => cliOutput.push({ stderr: data.toString() }));

        cli.on('close', code => {
          cliOutput.push({ code });

          generatedCode = fs.readFileSync(OUTPUT_PATH, 'utf-8');
          fs.unlinkSync(OUTPUT_PATH);

          resolve();
        });
      });
    });

    it('prints the expected stdout/stderr output', () => {
      expect(cliOutput).toMatchSnapshot();
    });

    it('generates the expected parser code', () => {
      expect(generatedCode).toMatchSnapshot();
    });
  });
});
