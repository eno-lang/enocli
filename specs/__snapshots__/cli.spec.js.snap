// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI generates a parser given valid arguments generates the expected parser code 1`] = `
"const enolib = require('enolib');
const fs = require('fs');

exports.parse = path => {
  const input = fs.readFileSync(path, 'utf-8');
  const _document = enolib.parse(input, { source: path });

  const document = {};
  document.field = _document.field('field').requiredStringValue();
  {
    const _fieldset = _document.requiredFieldset('fieldset');
    document.fieldset = {};
    document.fieldset.entry = _fieldset.entry('entry').requiredStringValue();
  }
  document.list = _document.requiredList('list').requiredStringValues();
  {
    document.section = {};
    const _section = _document.requiredSection('section');
    const section = document.section;
    section.field = _section.field('field').requiredStringValue();
  }

  return document;
};"
`;

exports[`CLI generates a parser given valid arguments prints the expected stdout/stderr output 1`] = `
Array [
  Object {
    "stdout": "Successfully generated javascript target '/home/simon/enocli/specs/temp_output/cli_output.js'.

Usage:
-----
const { parse } = require('/home/simon/enocli/specs/temp_output/cli_output.js'); // require path might need adjustments for your project
const obj = parse('samples/basic.eno');
",
  },
  Object {
    "code": 0,
  },
]
`;
